File: apps/api/app/schemas/news.py
"""Pydantic schemas for news resources."""
from __future__ import annotations

from datetime import datetime, date

from pydantic import Field, HttpUrl

from .base import CamelModel


class NewsSourceRead(CamelModel):
    id: str
    name: str
    url: HttpUrl
    gradient: tuple[str, str] = Field(..., description="Start and end hex colors")
    weight: float = Field(ge=0)


class GithubRepoRead(CamelModel):
    full_name: str
    stars: int = Field(ge=0)
    delta_stars: int


class NewsArticleRead(CamelModel):
    id: str
    title: str
    slug: str
    source_id: str
    url: HttpUrl
    summary: str
    tags: list[str] = Field(default_factory=list)
    published_at: datetime
    score: float
    github_repo: GithubRepoRead | None = None
    hero_image: HttpUrl | None = None
    content: list[str] = Field(default_factory=list)


class WeeklyDigestRead(CamelModel):
    week_of: date
    headline: str
    summary: str
    article_ids: list[str] = Field(default_factory=list, max_length=10)


class NewsPayload(CamelModel):
    sources: list[NewsSourceRead]
    articles: list[NewsArticleRead]
    digest: WeeklyDigestRead | None = None


class NewsArticleDetail(CamelModel):
    article: NewsArticleRead
    source: NewsSourceRead | None = None
    related: list[NewsArticleRead] = Field(default_factory=list, max_length=6)

File: packages/types/src/schemas/news.ts
import { z } from "zod";

export const newsSourceSchema = z.object({
  id: z.string(),
  name: z.string(),
  url: z.string().url(),
  gradient: z.tuple([z.string(), z.string()]),
  weight: z.number().min(0),
});

export type NewsSource = z.infer<typeof newsSourceSchema>;

export const newsArticleSchema = z.object({
  id: z.string(),
  title: z.string(),
  slug: z.string(),
  sourceId: z.string(),
  url: z.string().url(),
  summary: z.string(),
  tags: z.array(z.string()).default([]),
  publishedAt: z.string().datetime(),
  githubRepo: z
    .object({
      fullName: z.string(),
      stars: z.number().int().min(0),
      deltaStars: z.number().int(),
    })
    .optional(),
  heroImage: z.string().url().optional(),
  score: z.number(),
  content: z.array(z.string()).default([]),
});

export type NewsArticle = z.infer<typeof newsArticleSchema>;

export const weeklyDigestSchema = z.object({
  weekOf: z.string().date(),
  headline: z.string(),
  summary: z.string(),
  articleIds: z.array(z.string()).max(10),
});

export type WeeklyDigest = z.infer<typeof weeklyDigestSchema>;

export const newsPayloadSchema = z.object({
  sources: z.array(newsSourceSchema),
  articles: z.array(newsArticleSchema),
  digest: weeklyDigestSchema.nullable(),
});

export type NewsPayload = z.infer<typeof newsPayloadSchema>;

export const newsArticleDetailSchema = z.object({
  article: newsArticleSchema,
  source: newsSourceSchema.nullable(),
  related: z.array(newsArticleSchema),
});

export type NewsArticleDetail = z.infer<typeof newsArticleDetailSchema>;
